src/transformStateWithClones.js

```diff
@@ -7,7 +7,24 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      newState = {};
+    } else if (action.type === 'addProperties') {
+      newState = {
+        ...newState, ...action.extraData,
+      };
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach(key => delete newState[key]);
+    }
+
```

Comment:
checklist 3 [CODE STYLE]: case should always have default case for error handling.